### DISCUSSION TITLE
Unable to access request body in in api route

### DISCUSSION THREAD

**Discussion by Preet-Sojitra** [vercel/next.js#50607] [Help] [answered]
### Summary

I am using NextJs with pages directory and I am trying to access request's body in one api route but it is giving empty object. Here's the code:
```
export const config = {
  api: {
    bodyParser: true,
  },
}

export default function handler(req, res) {
  res.status(200).json(req.body)
}
```

Below is the image of the request I am sending via postman
![image](https://github.com/vercel/next.js/assets/87543808/566b33f1-7509-4ccc-aaf7-dddead62f7b3)

I have other api routes also and I tried accessing request body there and in that routes they are working perfectly. 

I am attaching the my github repo link. Refer the `backend` branch and code for this api route is in `/pages/api/login.js.`

### Example

https://github.com/Preet-Sojitra/preetsojitra.me/tree/backend

**Comment by icyJoseph** [comment]
Hi,

Are you using the header `content-type: application/json`? I just made a next.js pages dir app, and fired a request to it, worked without issues. Gonna check out your repo soon.

**Reply by icyJoseph** [reply, marked as answer]
Oh wait, you are using a GET request. That's the problem! You should use POST.

**Reply by Preet-Sojitra** [reply]
> Oh wait, you are using a GET request. That's the problem! You should use POST.

Can't we send body in `GET` request?

**Reply by icyJoseph** [reply]
Well, it is a touchy subject.

I'd say, if you can use POST, just do it.

The problem being that AFAIK the spec for a GET request is vague when it comes to body being allowed. And it has changed a few years ago too.

The key thing here is that different products will have different behavior. 

Why does it become touchy for developers? Because from the Command Query Separation, principle, a GET is a query and it shouldn't mutate stuff, like doing a user login. It should be idempotent.

Many can argue that the body might be used to carry parameters for the query, but then again we have query parameters in the URL.

There's also CDN caching which AFAIK, doesn't take GET body into consideration. Only headers and the URL. CDNs might cache GET requests by default, and skip caching POST requests by default too. 

I tried to manually parse the body out of the request anyway, but got empty string too. Again, I'd say, take POST, clearly GET requests w/ body are not working. 

I read that it took about 3 years to convince the Postman team to allow sending a body with GET requests. So it's already a big deal that you can send the request to begin with. 

I'll check later today what's the behavior on the new Route Handlers in the App Directory. Would be funny if it worked there 😅

**Reply by aelsaman** [reply]
Thanks for clarifying that @icyJoseph . It is very odd how Vercel is simply ignore all incoming GET requests that has a body attached as if it is not being sent (without returning errors/warnings whatsoever). I spent sometime trying to figure it out and found zero clues/hints on their documentation related to using POST request instead of GET request in case of a body being sent.
